core/dbt/exceptions.py

```diff
@@ -209,6 +209,24 @@ def _fix_dupe_msg(self, path_1: str, path_2: str, name: str, type_name: str) ->
             )
 
 
+class ModelContractError(DbtRuntimeError):
+    CODE = 10016
+    MESSAGE = "Contract Error"
+
+    def __init__(self, node=None):
+        super().__init__(self.message(), node)
+
+    @property
+    def type(self):
+        return "Contract"
+
+    def message(self):
+        return (
+            "There is a breaking change in the model contract; "
+            "you may need to create a new version. See: https://docs.getdbt.com/docs/collaborate/publish/model-versions"
```

Comment:
This error message looks good to me 👍

Reply:
If the user is aware they're making a breaking change, and just needs some guidance regarding making a new version of the model - this error message looks good to me too. 

That said, I am worried that this error message could be a little opaque for unintented breaking changes, especially as [our set of conditions that constitute breaking changes becomes more nuanced](https://github.com/dbt-labs/dbt-core/issues/7065). For example, a user updates a contracted model with constraints from `table` to `view` in an attempt to optimize some spend, which raises a breaking change error because the existing constraints can't be validated. Just seeing `There is a breaking change in the model contract` could be confusing, and perhaps the user would actually choose to just revert that change and make it at a later point (bundle it with some other breaking changes). In that scenario, I'd (as an imaginary user) love to see an error message that looks more like: 
```
There is a breaking change in the model contract: 
 * Updated materialization from `table` to `view` for model with constraints
 * Updated config from contract: true to contract: false
 * Column a removed
 * Column c updated data_type from string to int
 * ...
You may need to create a new version. See: https://docs.getdbt.com/docs/collaborate/publish/model-versions
```

This may not be concern with the current set of conditions for breaking changes (any change to column name/data_type => breaking change), but given that we're planning to extend this to be more nuanced soon in https://github.com/dbt-labs/dbt-core/issues/7065, I'd leave some design space for it if this scenario seems important to prioritize the UX of (cc @jtcohen6).

Reply:
Currently the change of a materialization is not a contract change, it's a config change. The ticket doesn't mention materialization -- do you want that to be part of the contract change too?

Creating that more user-friendly error message will probably be about twice as much work than everything else so far. If you want me to spend my time doing that, I can.

Reply:
Materializations are not part of this issue, but they will be part the work to extend detecting breaking changes in https://github.com/dbt-labs/dbt-core/issues/7065 (which i still blocked by https://github.com/dbt-labs/dbt-core/issues/7067). If we know we'll want more granular error messaging as part of that work, we could set up the foundations here or leave it to https://github.com/dbt-labs/dbt-core/issues/7065.

Reply:
I added a brief "reason" to the exception message. We can do more elaborate messaging once the other parts are in place.

Reply:
(good conversation, no notes!)